File: lint-rules-android-lint/src/main/java/com/vanniktech/lintrules/android/SuperfluousPaddingDeclarationDetector.java
package com.vanniktech.lintrules.android;

import com.android.tools.lint.detector.api.Implementation;
import com.android.tools.lint.detector.api.Issue;
import java.util.Collection;

import static com.android.SdkConstants.ATTR_PADDING_BOTTOM;
import static com.android.SdkConstants.ATTR_PADDING_END;
import static com.android.SdkConstants.ATTR_PADDING_START;
import static com.android.SdkConstants.ATTR_PADDING_TOP;
import static com.android.tools.lint.detector.api.Category.CORRECTNESS;
import static com.android.tools.lint.detector.api.Scope.RESOURCE_FILE_SCOPE;
import static com.android.tools.lint.detector.api.Severity.WARNING;
import static java.util.Arrays.asList;

public final class SuperfluousPaddingDeclarationDetector extends SuperfluousDeclarationDetector {
  static final Issue ISSUE_SUPERFLUOUS_PADDING_DECLARATION = Issue.create("SuperfluousPaddingDeclaration", "Instead of using start, end, bottom and top padding can be used.",
      "Instead of using start, end, bottom and top padding can be used.", CORRECTNESS, 8, WARNING,
      new Implementation(SuperfluousPaddingDeclarationDetector.class, RESOURCE_FILE_SCOPE));

  @Override public Collection<String> getApplicableAttributes() {
    return asList(
        ATTR_PADDING_TOP,
        ATTR_PADDING_BOTTOM,
        ATTR_PADDING_START,
        ATTR_PADDING_END
    );
  }

  @Override public Issue issue() {
    return ISSUE_SUPERFLUOUS_PADDING_DECLARATION;
  }

  @Override public String message() {
    return "Should be using padding instead.";
  }
}

File: lint-rules-android-lint/src/main/kotlin/com/vanniktech/lintrules/android/SuperfluousPaddingDeclarationDetector.kt
@file:Suppress("UnstableApiUsage") // We know that Lint APIs aren't final.

package com.vanniktech.lintrules.android

import com.android.SdkConstants.ATTR_PADDING_BOTTOM
import com.android.SdkConstants.ATTR_PADDING_END
import com.android.SdkConstants.ATTR_PADDING_START
import com.android.SdkConstants.ATTR_PADDING_TOP
import com.android.tools.lint.detector.api.Category.Companion.CORRECTNESS
import com.android.tools.lint.detector.api.Implementation
import com.android.tools.lint.detector.api.Issue
import com.android.tools.lint.detector.api.Scope.Companion.RESOURCE_FILE_SCOPE
import com.android.tools.lint.detector.api.Severity.WARNING
import java.util.Arrays.asList

val ISSUE_SUPERFLUOUS_PADDING_DECLARATION = Issue.create(
  "SuperfluousPaddingDeclaration",
  "Flags padding declarations that can be simplified.",
  "Instead of using start-, end-, bottom- and top paddings, padding can be used.",
  CORRECTNESS, PRIORITY, WARNING,
  Implementation(SuperfluousPaddingDeclarationDetector::class.java, RESOURCE_FILE_SCOPE),
)

class SuperfluousPaddingDeclarationDetector : SuperfluousDeclarationDetector(
  applicableSuperfluousAttributes = asList(
    ATTR_PADDING_TOP,
    ATTR_PADDING_BOTTOM,
    ATTR_PADDING_START,
    ATTR_PADDING_END,
  ),
  issue = ISSUE_SUPERFLUOUS_PADDING_DECLARATION,
  message = "Should be using padding instead.",
)
